hr_timesheet_sheet/models/hr_timesheet_sheet.py

```diff
@@ -0,0 +1,621 @@
+        for rec in self:
+            if rec.state in ['new', 'draft']:
+                rec.add_line()
+                rec.add_line_task_id = False
+                rec.add_line_project_id = False
+        return True
+
+    def _get_date_name(self, date):
+        return fields.Date.from_string(date).strftime("%a\n%b %d")
+
+    def _get_dates(self):
+        start = fields.Date.from_string(self.date_start)
+        end = fields.Date.from_string(self.date_end)
+        if end < start:
+            return []
+        # time_period = end - start
+        # number_of_days = time_period/timedelta(days=1)
+        dates = [fields.Date.to_string(start)]
+        while start != end:
+            start += relativedelta(days=1)
+            dates.append(fields.Date.to_string(start))
+        return dates
+
+    def _get_line_name(self, project, task=None):
+        name = '{}'.format(project.name)
+        if task:
+            name += ' - {}'.format(task.name)
+        return name
+
+    def _get_default_analytic_line(self, date, project, task, timesheet=None):
+        name_y = self._get_line_name(project, task)
+        self.clean_timesheets(timesheet)
+        values = {
+            'value_x': self._get_date_name(date),
+            'value_y': name_y,
+            'date': date,
+            'project_id': project.id,
+            'task_id': task and task.id or False,
+            'count_timesheets': len(timesheet),
+            'unit_amount': 0.0,
+        }
+        if self.id:
+            values.update({
+                'sheet_id': self.id,
+            })
+        if timesheet:
+            amount = sum([t.unit_amount for t in timesheet])
+            values.update({
+                'unit_amount': amount,
+            })
+        return values
+
+    @api.model
+    def _prepare_empty_analytic_line(self):
+        return {
+            'name': '/',
+            'employee_id': self.employee_id.id,
+            'date': self.date_start,
+            'project_id': self.add_line_project_id and
+            self.add_line_project_id.id or False,
+            'task_id': self.add_line_task_id and
+            self.add_line_task_id.id or False,
+            'sheet_id': self.id,
+            'unit_amount': 0.0,
+            'count_timesheets': 0,
+        }
+
+    @api.model
+    def add_line(self):
+        if self.add_line_project_id:
+            values = self._prepare_empty_analytic_line()
+            name_lines = self.line_ids.mapped('value_y')
+            name_line = self._get_line_name(
+                self.add_line_project_id, self.add_line_task_id)
+            if name_lines:
+                self.delete_empty_lines(name_lines)
+            if name_line not in name_lines:
+                self.timesheet_ids |= \
+                    self.env['account.analytic.line'].create(values)
+
+    def clean_timesheets(self, timesheet):
+        if self.id:
+            for anal_line in timesheet.filtered(lambda a: not a.sheet_id):
```

Comment:
:fire: can we change the spelling here? :smile:

Reply:
hahaha

Reply:
@lreficent 😂

Reply:
[here](https://github.com/OCA/account-invoicing/blob/10.0/account_invoice_merge/models/account_invoice.py#L207) nobody was burned 😜